restui/views/mappings.py

```diff
@@ -270,39 +268,37 @@ def put(self, request, pk):
         except MultipleObjectsReturned:
             raise Http404("Couldn't get unique status for {}".format(request.data['status']))
 
-#        for old_status in UeMappingStatus.objects.filter(mapping=mapping).filter(~Q(status=s.id)):
-#            old_status.delete()
-
         # If the mapping has already been assigned that status, update the timestamp,
         # otherwise create one from scratch
         try:
-#            mapping_status = UeMappingStatus.objects.get(mapping=mapping, status=s.id)
-            mapping_status = UeMappingStatus.objects.get(mapping=mapping)
+            mapping_status = UeMappingStatus.objects.filter(mapping=mapping).latest('time_stamp')
         except UeMappingStatus.DoesNotExist:
-            # create new mapping status
-            serializer = StatusSerializer(data={ 'time_stamp':timezone.now(),
-                                                 'user_stamp':request.user,
-                                                 'status':s.id,
-                                                 'mapping':mapping.mapping_id })
+            # It's alright, for the first status change of a mapping a
+            # historic record won't exist.
+            pass
 
-            if serializer.is_valid():
-                serializer.save()
-                return Response(serializer.data, status=status.HTTP_201_CREATED)
+        else:
+            if mapping_status.status == s.id:
+                # The user is trying to change it to what the status
+                # already is, nothing to do.
+                return Response(status=status.HTTP_204_NO_CONTENT)
 
-        except MultipleObjectsReturned:
-            raise Http404("Couldn't get unique status for mapping {}".format(mapping.mapping_id))
+        # create new mapping status
+        serializer = StatusSerializer(data={ 'time_stamp':timezone.now(),
+                                             'user_stamp':request.user,
+                                             'status':s.id,
+                                             'mapping':mapping.mapping_id })
+
+        if serializer.is_valid():
+            serializer.save()
         else:
-            # mapping status already exist, update timestamp
-            serializer = StatusSerializer(mapping_status, data={ 'time_stamp':timezone.now(),
-                                                                 'user_stamp':request.user,
-                                                                 'status':s.id,
-                                                                 'mapping':mapping.mapping_id })
-            
-            if serializer.is_valid():
-                serializer.save()
-                return Response(serializer.data)
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
 
-        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+        # Update the status in the mapping record
```

Comment:
But we also need to update the status history with the latest status. Or am I missing something?

Reply:
That's what's happening in the try block. It tries to add the new history, but if the new history entry is the same status, nothing changes and return the 204. If we make it this far we've already added the new history record up on line 293.